pkg/controllers/updaterun/stop.go

```diff
@@ -0,0 +1,237 @@
+
+	placementv1beta1 "github.com/kubefleet-dev/kubefleet/apis/placement/v1beta1"
+	"github.com/kubefleet-dev/kubefleet/pkg/utils/condition"
+	"github.com/kubefleet-dev/kubefleet/pkg/utils/controller"
+)
+
+// stop handles stopping the update run.
+func (r *Reconciler) stop(
+	updateRun placementv1beta1.UpdateRunObj,
+	updatingStageIndex int,
+	toBeUpdatedBindings, toBeDeletedBindings []placementv1beta1.BindingObj,
+) (finished bool, waitTime time.Duration, stopErr error) {
+	updateRunStatus := updateRun.GetUpdateRunStatus()
+	var updatingStageStatus *placementv1beta1.StageUpdatingStatus
+
+	// Set up defer function to handle errStagedUpdatedAborted.
+	defer func() {
+		checkIfErrorStagedUpdateAborted(stopErr, updateRun, updatingStageStatus)
+	}()
+
+	markUpdateRunStopping(updateRun)
+
+	if updatingStageIndex < len(updateRunStatus.StagesStatus) {
+		return r.stopUpdatingStage(updateRun, updatingStageIndex, toBeUpdatedBindings)
+	}
+	// All the stages have finished, stop the delete stage.
+	finished, stopErr = r.stopDeleteStage(updateRun, toBeDeletedBindings)
+	return finished, clusterUpdatingWaitTime, stopErr
+}
+
+// stopUpdatingStage stops the updating stage by letting the updating bindings finish and not starting new updates.
+func (r *Reconciler) stopUpdatingStage(
+	updateRun placementv1beta1.UpdateRunObj,
+	updatingStageIndex int,
+	toBeUpdatedBindings []placementv1beta1.BindingObj,
+) (bool, time.Duration, error) {
+	updateRunStatus := updateRun.GetUpdateRunStatus()
+	updatingStageStatus := &updateRunStatus.StagesStatus[updatingStageIndex]
+	updateRunRef := klog.KObj(updateRun)
+	// Create the map of the toBeUpdatedBindings.
+	toBeUpdatedBindingsMap := make(map[string]placementv1beta1.BindingObj, len(toBeUpdatedBindings))
+	for _, binding := range toBeUpdatedBindings {
+		bindingSpec := binding.GetBindingSpec()
+		toBeUpdatedBindingsMap[bindingSpec.TargetCluster] = binding
+	}
+	// Mark the stage as stopping in case it's not.
+	markStageUpdatingStopping(updatingStageStatus, updateRun.GetGeneration())
+	clusterUpdatingCount := 0
+	var stuckClusterNames []string
+	var clusterUpdateErrors []error
+	// Go through each cluster in the stage and check if it's updating/succeeded/failed/not started.
+	for i := 0; i < len(updatingStageStatus.Clusters); i++ {
+		clusterStatus := &updatingStageStatus.Clusters[i]
+		clusterStartedCond := meta.FindStatusCondition(clusterStatus.Conditions, string(placementv1beta1.ClusterUpdatingConditionStarted))
+		if !condition.IsConditionStatusTrue(clusterStartedCond, updateRun.GetGeneration()) {
+			// Cluster has not started updating therefore no need to do anything.
+			continue
+		}
+
+		clusterUpdateSucceededCond := meta.FindStatusCondition(clusterStatus.Conditions, string(placementv1beta1.ClusterUpdatingConditionSucceeded))
+		if condition.IsConditionStatusFalse(clusterUpdateSucceededCond, updateRun.GetGeneration()) || condition.IsConditionStatusTrue(clusterUpdateSucceededCond, updateRun.GetGeneration()) {
```

Comment:
If an update to a cluster fails, should we abort here ?.

Reply:
We can continue if a cluster update fails, like our release pipeline, though we currently does not have a very confident way to say cluster update has failed. We currently only fail when the cluster is preempted by another updateRun but that would abort the updateRun during execution.

Reply:
basically during stop I think we don't need to worry about a cluster that is already in a terminal state.